File: 202502/20250226-1749. Maximum Absolute Sum of Any Subarray/idea2.cpp
#include <iostream>
#include <vector>
using namespace std;
class Solution {
public:
    int maxAbsoluteSum(vector<int>& nums) {
        int maxabsSum = 0;
        int imin =  0; //minimum sum value for subarrays end with index i
        int imax = 0; //maximum sum value for subarrays end with index i
        for (int val : nums){
            imin = imin<0? imin+val:val;
            imax = imax>0? imax+val:val;
            maxabsSum = max({maxabsSum, abs(imin), abs(imax)});
        }
        return maxabsSum;
    }
};

int main(int argc, char const *argv[]){
    vector<int> nums = {1,-3,2,3,-2};
    Solution s;
    cout << s.maxAbsoluteSum(nums) << endl;
    return 0;
}

File: 202504/07-416. Partition Equal Subset Sum/idea2.py
from typing import *

class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        N = len(nums)
        total = sum(nums)
        if total%2 != 0:
            return False
        else:
            half = total//2
        dp = [[0]*(half+1) for x in range(N)]
        for i in range(1, N):
            num = nums[i-1]
            if num > half:
                return False
            for j in range(1, half+1):
                if num > j:
                    dp[i][j] = dp[i-1][j]
                else:
                    dp[i][j] = max(
                        dp[i-1][j],
                        dp[i-1][j-num]+num
                    )
            if dp[i][-1] == half:
                return True
        return False
                    
    
    
s = Solution()
# print(s.canPartition([1,5,11,5]))
# print(s.canPartition([3,3,3,4,5]))
print(s.canPartition([1,2,3,4,5,6,7]))